### ISSUE TITLE
smart patching (right/click patching)

### ISSUE THREAD

**Issue by psifertex** [Vector35/binaryninja-api#27]
Ability to appropriately resize or find 'dead' space, insert jumps, copy existing code, integrate patches with existing relative offsets and symbols and then add new functionality to patch on the fly.

Needs support for: 
- [ ] Mach-o
- [ ] ELF
- [ ] PE

**Comment by CryptXor** [comment]
Could you add an option to "preview" the changes before they are added to that thought.

**Comment by plafosse** [comment]
Since you can use undo do you feel a preview is really necessary?

**Comment by CryptXor** [comment]
Yes, as it would be easy to see if there are any bugs with over-writing/inserting.
Perhaps such options could be added as "debug" options enabled via the prefs?

**Comment by psifertex** [comment]
Note that BNHook implements this functionality as a plugin for ELF: https://github.com/orndorffgrant/bnhook

**Comment by xusheng6** [comment]
Just wish to update this that now that we have the ability to create a new segment not backed by file and write to it, it seems better to just write to there rather than trying to find empty spaces in the executable. That said, my suggested approach would only do good for analysis, if one wants the code to execute the patch, we still need to change the section/segment headers for it
